src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using FilterResult = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.FilterResult;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using Session = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class ItemManager : ForegroundThreadAffinitizedObject, IAsyncCompletionItemManager
+    {
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        internal ItemManager(IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<VSCompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        public Task<AsyncCompletionData.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+            => Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+
+        private AsyncCompletionData.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            if (!session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default hasSuggestedItemOptions is false. We can discuss if the opposite is required.
+                hasSuggestedItemOptions = false;
+            }
```

Comment:
if the session is created outside of roslyn... why are we doing anything?

Reply:
We were asked to support those scenarios. Please see other comments.

---
In reply to: [232509596](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509596) [](ancestors = 232509596)